Allow signs to be 0 width

lukas-reineke opened the pull request:
Currently, signs need to be 1 or 2 cell wide.
This PR would allow them to be 0, and don't display the signcolumn for
those signs. So you can use only `linehl` or `numhl` for _some_ signs. 

![Screen Capture_select-area_20201114104128](https://user-images.githubusercontent.com/12900252/99136229-29d14680-2667-11eb-9d47-48718b744c98.png)

![Screen Capture_select-area_20201114104209](https://user-images.githubusercontent.com/12900252/99136319-8cc2dd80-2667-11eb-9eac-9f9660fe25cd.png)

What do you think about the idea?
Code probably needs cleanup + tests etc.

lukas-reineke commented:
fixed the tests + added some new ones

janlazo commented on src/nvim/testdir/test_signs.vim:
```diff
@@ -133,9 +133,8 @@ func Test_sign()
-  " Only 1 or 2 character text is allowed
+  " Only 0, 1 or 2 character text is allowed
   call assert_fails("sign define Sign4 text=abc linehl=Comment", 'E239:')
-  call assert_fails("sign define Sign4 text= linehl=Comment", 'E239:')
```
It's fine to update comments to reflect new behavior but line deletions should be avoided in legacy tests.

lukas-reineke replied:
But this will not fail anymore.
I could add a test instead to assert that it doesn't fail, but I already did this in the lua test.

janlazo replied:
Comment it out.

lukas-reineke replied:
Ah I see what you mean. Updated

lukas-reineke committed:
Allow signs to be 0 width

Adds support for signs to be 0 cells wide. If all signs of the same
group have no width, the signcolumn will not be rendered for that group.

janlazo reviewed (approved):
LGTM.

Can you investigate if any unmerged Vim patches conflict with this PR?

lukas-reineke commented:
> Can you investigate if any unmerged Vim patches conflict with this PR?

I don't see any open patch that touches `sign.c` or `buffer.c`. I think it should be fine.

teto commented:
I am all ok with the idea a priori, I would like to test it first, this WE probably. Have you inquired if vim folks were interested ?

lukas-reineke commented:
> Have you inquired if vim folks were interested ?

No I haven't. But the implementation of how the signcolumn is computeted is different in Vim.

teto commented:
merry christmas
